lib/common/rules.c

```diff
@@ -1155,41 +1155,66 @@ pcmk__evaluate_attr_expression(const xmlNode *expression,
  * \internal
  * \brief Evaluate a resource rule expression
  *
- * \param[in] expr        XML of a rule's \c PCMK_XE_RSC_EXPRESSION subelement
- * \param[in] rule_input  Values used to evaluate rule criteria
+ * \param[in] rsc_expression  XML of rule's \c PCMK_XE_RSC_EXPRESSION subelement
+ * \param[in] rule_input      Values used to evaluate rule criteria
  *
  * \return Standard Pacemaker return code (\c pcmk_rc_ok if the expression
  *         passes, some other value if it does not)
  */
 int
-pcmk__evaluate_rsc_expression(const xmlNode *expr,
+pcmk__evaluate_rsc_expression(const xmlNode *rsc_expression,
                               const pcmk_rule_input_t *rule_input)
 {
-    const char *class = crm_element_value(expr, PCMK_XA_CLASS);
-    const char *provider = crm_element_value(expr, PCMK_XA_PROVIDER);
-    const char *type = crm_element_value(expr, PCMK_XA_TYPE);
+    const char *id = NULL;
+    const char *value = NULL;
 
-    crm_trace("Testing rsc_defaults expression: %s", pcmk__xe_id(expr));
+    if ((rsc_expression == NULL) || (rule_input == NULL)) {
+        return EINVAL;
+    }
+
+    // Validate XML ID
+    id = pcmk__xe_id(rsc_expression);
+    if (pcmk__str_empty(id)) {
+        // Not possible with schema validation enabled
+        /* @COMPAT When we can break behavioral backward compatibility,
+         * fail the expression
+         */
+        pcmk__config_warn(PCMK_XE_RSC_EXPRESSION " has no " PCMK_XA_ID);
+        id = "without ID"; // for logging
+    }
 
-    if (class != NULL &&
-        !pcmk__str_eq(class, rule_input->rsc_standard, pcmk__str_none)) {
-        crm_trace("Class doesn't match: %s != %s", class, rule_input->rsc_standard);
+    // Compare resource standard
+    value = crm_element_value(rsc_expression, PCMK_XA_CLASS);
```

Comment:
Similar slight behavioral change to the one I mentioned for operation expressions:
Previously, `PCMK_XA_CLASS == "", rule_input->rsc_standard == "ocf"` would fail. Now, it passes. Same for provider and type.

Caveats similar to the ones I mentioned before also apply.

Reply:
Thx